source/s3_client.c

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "s3_client.h"
+
+#include "auth.h"
+#include "io.h"
+
+static const char *s_capsule_name_s3_client = "aws_s3_client";
+
+struct s3_client_binding {
+    struct aws_s3_client *native;
+
+    bool release_called;
+    bool shutdown_called;
+
+    /* Shutdown callback, reference cleared after invoking callback */
+    PyObject *on_shutdown;
+};
+
+static void s_destroy_if_ready(struct s3_client_binding *client) {
+    if (client->native && (!client->shutdown_called || !client->release_called)) {
+        /* native client successfully created, but not ready to clean up yet */
+        return;
+    }
+    /* in case native never existed and shutdown never happened */
+    Py_XDECREF(client->on_shutdown);
+    aws_mem_release(aws_py_get_allocator(), client);
```

Comment:
trivial: just delete the "in case native never existed" comment. this code path is also taken if native DID exist and normal shutdown is complete. The comment above this, about "not ready to clean up yet" conveys this all pretty well